bittide/src/Bittide/ClockControl/ClockGenConfig.hs

```diff
@@ -0,0 +1,274 @@
+  , Signal dom Bool
+  , ( Signal dom Bool -- SCK
+    , Signal dom Bit  -- MOSI
+    , Signal dom Bool -- SS
+    )
+  )
+si539xSpi ClockGenRegisterMap{..} minTargetPs@SNat externalOperation miso =
+  (configByte, configBusy, configSuccess, spiOut)
+ where
+  (driverByte, driverBusy, spiOut) = si539xSpiDriver minTargetPs spiOperation miso
+  romOut = rom (configPreamble ++ config ++ configPostamble) $ bitCoerce <$> readCounter
+
+  readCounter :: Signal dom (Index (preambleEntries + configEntries + postambleEntries))
+  (readCounter, spiOperation, configBusy, configByte, configSuccess) =
+    mealyB go (FetchReg 0) (romOut, externalOperation, driverByte, driverBusy)
+
+  go currentState ((regPage,regAddress,byte), extSpi, spiByte, spiBusy) =
+    (nextState, (getStateAddress currentState, spiOp, busy, returnedByte, currentState == Finished))
+   where
+    isConfigEntry i = (natToNum @preambleEntries) <= i && i < (natToNum @(preambleEntries + configEntries))
+    nextState = case (currentState, spiByte) of
+      (WriteEntry i, Just _)
+        | i == maxBound -> Finished
+        | i == (natToNum @preambleEntries - 1) -> Wait (0 :: Index (PeriodCycles dom (300*10^9))) i
+        | isConfigEntry i -> ReadEntry i
+        | otherwise -> FetchReg (succ i)
+
+      (ReadEntry i, Just b)
+        | b == byte -> FetchReg (succ i)
+        | otherwise -> Error i
+
+      (FetchReg i, _)                   -> WriteEntry i
+      (Wait ((==maxBound) -> True) i, _)-> FetchReg (succ i)
+      (Wait j i, _)                     -> Wait (succ j) i
+      (WriteEntry _, Nothing)           -> currentState
+      (ReadEntry _, Nothing)            -> currentState
+      (Finished , _)                    -> currentState
+      (Error _ , _)                     -> currentState
+
+    spiOp = case currentState of
+      FetchReg _   -> Nothing
+      WriteEntry _ -> Just RegisterOperation{regPage, regAddress, regWrite = Just byte}
+      ReadEntry _  -> Just RegisterOperation{regPage, regAddress, regWrite = Nothing}
+      Wait _ _     -> Nothing
+      Finished     -> extSpi
+      Error _      -> extSpi
+
+    (busy, returnedByte) = case currentState of
+      Finished -> (spiBusy, spiByte)
+      Error _  -> (spiBusy, spiByte)
+      _        -> (True, Nothing)
+
+-- | Keeps track of the current 'Page' and 'Address' of the @Si539x@ chip.
+-- Furthermore is used to track the current communication cycle.
+data DriverState dom = DriverState
+  { currentPage         :: Maybe Page
+  -- ^ Current 'Page' of the @Si539x@ chip.
+  , currentAddress      :: Maybe Address
+  -- ^ Current 'Address' of the @Si539x@ chip.
+  , currentOp           :: Maybe RegisterOperation
+  -- ^ Current communication transaction.
+  , commandAcknowledged :: Acknowledge
+  -- ^ Whether or not the current transaction has already been acknowledged.
+  , idleCycles          :: Index (PeriodCycles dom 95000)
+  -- ^ After communication, slave select must be high for at least 95ns.
```

Comment:
```suggestion
  , idleCycles          :: Index (PeriodCycles dom (95 * Ns))
  -- ^ After communication, slave select must be high for at least 95ns.
```

Gergo Erdi defines some pretty handy type aliases, we should probably steal them.